No Matching Netty Transport after upgrade to Micronaut 4.10.0

### Expected Behavior

After upgrading to Micronaut 4.10.0, a Micronaut HTTP service should start without error.

### Actual Behaviour

One of our services fails to start due to the following error:

> Caused by: java.util.NoSuchElementException: No matching transport was found. Configured transports are [nio], available transports are [instrumentedNioEventLoopGroupFactory, epoll, native]

This only occurs in one of our services.  All services are containerized and none of them contain any Micronaut configuration to change/modify the Netty event loop configuration.  It is also worth noting that we can only reproduce this in one of our clusters.

### Steps To Reproduce

1. Update to Micronaut 4.10.0 (from 4.9.x)
2. Start the service
3. Service fails on startup due to the `No matching transport was found. Configured transports are [nio], available transports are [instrumentedNioEventLoopGroupFactory, epoll, native]` error.

### Environment Information

 Service running in Docker container (airbyte-workload-api-service)

### Example Application

https://github.com/airbytehq/airbyte-platform

### Version

4.10.0

--- a/http-netty/src/main/java/io/micronaut/http/netty/channel/DefaultEventLoopGroupRegistry.java
+++ b/http-netty/src/main/java/io/micronaut/http/netty/channel/DefaultEventLoopGroupRegistry.java
@@ -30,8 +30,10 @@ import io.micronaut.core.util.ArgumentUtils;
 import io.micronaut.http.netty.channel.loom.LoomCarrierGroup;
 import io.micronaut.inject.qualifiers.Qualifiers;
 import io.netty.channel.EventLoopGroup;
+import io.netty.channel.IoEventLoop;
 import io.netty.channel.IoHandlerFactory;
 import io.netty.channel.MultiThreadIoEventLoopGroup;
+import io.netty.channel.SingleThreadIoEventLoop;
 import io.netty.channel.nio.NioEventLoopGroup;
 import io.netty.util.NettyRuntime;
 import io.netty.util.concurrent.DefaultThreadFactory;
@@ -42,8 +44,10 @@ import jakarta.inject.Singleton;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.util.List;
 import java.util.Map;
 import java.util.Optional;
+import java.util.Queue;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.Executor;
 import java.util.concurrent.ThreadFactory;
@@ -68,6 +72,8 @@ public class DefaultEventLoopGroupRegistry implements EventLoopGroupRegistry {
 
     private final BeanProvider<LoomCarrierGroup.Factory> loomCarrierGroupFactory;
 
+    private final List<TaskQueueInterceptor> taskQueueInterceptors;
+
     /**
      * Default constructor.
      *
@@ -75,10 +81,11 @@ public class DefaultEventLoopGroupRegistry implements EventLoopGroupRegistry {
      * @param beanLocator           The bean locator
      * @param loomCarrierGroupFactory Factory for the loom carrier group
      */
-    public DefaultEventLoopGroupRegistry(EventLoopGroupFactory eventLoopGroupFactory, BeanLocator beanLocator, BeanProvider<LoomCarrierGroup.Factory> loomCarrierGroupFactory) {
+    public DefaultEventLoopGroupRegistry(EventLoopGroupFactory eventLoopGroupFactory, BeanLocator beanLocator, BeanProvider<LoomCarrierGroup.Factory> loomCarrierGroupFactory, List<TaskQueueInterceptor> taskQueueInterceptors) {
         this.eventLoopGroupFactory = eventLoopGroupFactory;
         this.beanLocator = beanLocator;
         this.loomCarrierGroupFactory = loomCarrierGroupFactory;
+        this.taskQueueInterceptors = taskQueueInterceptors;
     }
 
     /**
@@ -100,14 +107,30 @@ public class DefaultEventLoopGroupRegistry implements EventLoopGroupRegistry {
         eventLoopGroups.clear();
     }
 
-    private EventLoopGroup createGroup(EventLoopGroupConfiguration configuration, Executor executor) {
+    private EventLoopGroup createGroup(EventLoopGroupConfiguration configuration, String name, Executor executor) {
         IoHandlerFactory ioHandlerFactory = eventLoopGroupFactory.createIoHandlerFactory(configuration);
         int nThreads = numThreads(configuration);
         EventLoopGroup eventLoopGroup;
         if (configuration.isLoomCarrier()) {
             eventLoopGroup = loomCarrierGroupFactory.get().create(nThreads, executor, ioHandlerFactory);
-        } else {
+        } else if (taskQueueInterceptors.isEmpty()) {
             eventLoopGroup = new MultiThreadIoEventLoopGroup(nThreads, executor, ioHandlerFactory);
+        } else {
+            eventLoopGroup = new MultiThreadIoEventLoopGroup(nThreads, executor, ioHandlerFactory) {
+                @Override
+                protected IoEventLoop newChild(Executor executor, IoHandlerFactory ioHandlerFactory, Object... args) {
+                    return new SingleThreadIoEventLoop(this, executor, ioHandlerFactory) {
+                        @Override
+                        protected Queue<Runnable> newTaskQueue(int maxPendingTasks) {
+                            Queue<Runnable> tq = super.newTaskQueue(maxPendingTasks);
+                            for (TaskQueueInterceptor taskQueueInterceptor : taskQueueInterceptors) {
+                                tq = taskQueueInterceptor.wrapTaskQueue(name, tq);
+                            }
+                            return tq;
+                        }
+                    };
+                }
+            };
         }
         eventLoopGroups.put(eventLoopGroup, configuration);
         return eventLoopGroup;
@@ -137,7 +160,7 @@ public class DefaultEventLoopGroupRegistry implements EventLoopGroupRegistry {
             executor = new ThreadPerTaskExecutor(threadFactory);
         }
 
-        return createGroup(configuration, executor);
+        return createGroup(configuration, configuration.getName(), executor);
     }
 
     /**
@@ -155,7 +178,7 @@ public class DefaultEventLoopGroupRegistry implements EventLoopGroupRegistry {
         if (threadFactory instanceof NettyThreadFactory.EventLoopCustomizableThreadFactory custom) {
             threadFactory = custom.customizeForEventLoop();
         }
-        return createGroup(new DefaultEventLoopGroupConfiguration(), new ThreadPerTaskExecutor(threadFactory));
+        return createGroup(new DefaultEventLoopGroupConfiguration(), EventLoopGroupConfiguration.DEFAULT, new ThreadPerTaskExecutor(threadFactory));
     }
 
     @NonNull

--- a/http-netty/src/main/java/io/micronaut/http/netty/channel/TaskQueueInterceptor.java
+++ b/http-netty/src/main/java/io/micronaut/http/netty/channel/TaskQueueInterceptor.java
@@ -1,0 +1,41 @@
+/*
+ * Copyright 2017-2025 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.http.netty.channel;
+
+import io.micronaut.context.annotation.BootstrapContextCompatible;
+import io.micronaut.core.annotation.Experimental;
+import io.micronaut.core.annotation.NonNull;
+
+import java.util.Queue;
+
+/**
+ * An interceptor that wraps a task queue for a netty event loop.
+ *
+ * @author Jonas Konrad
+ */
+@BootstrapContextCompatible
+@Experimental
+public interface TaskQueueInterceptor {
+    /**
+     * Wrap a task queue.
+     *
+     * @param groupName The event loop group name
+     * @param original  The original queue
+     * @return The wrapped queue
+     */
+    @NonNull
+    Queue<Runnable> wrapTaskQueue(@NonNull String groupName, @NonNull Queue<Runnable> original);
+}

--- a/http-netty/src/test/java/io/micronaut/http/netty/channel/TaskQueueInterceptorTest.java
+++ b/http-netty/src/test/java/io/micronaut/http/netty/channel/TaskQueueInterceptorTest.java
@@ -1,0 +1,73 @@
+package io.micronaut.http.netty.channel;
+
+import io.micronaut.context.ApplicationContext;
+import io.micronaut.context.annotation.Requires;
+import io.netty.channel.EventLoop;
+import io.netty.channel.EventLoopGroup;
+import io.netty.util.concurrent.Promise;
+import jakarta.inject.Singleton;
+import org.junit.jupiter.api.Test;
+
+import java.util.AbstractQueue;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class TaskQueueInterceptorTest {
+    @Test
+    public void test() throws ExecutionException, InterruptedException {
+        try (ApplicationContext ctx = ApplicationContext.run(Map.of("spec.name", "TaskQueueInterceptorTest"))) {
+            AtomicInteger started = ctx.getBean(MyTaskQueueInterceptor.class).started;
+            EventLoopGroup group = ctx.getBean(EventLoopGroup.class);
+            EventLoop loop = group.next();
+            Promise<String> promise = loop.newPromise();
+            int before = started.get();
+            loop.execute(() -> promise.setSuccess("foo"));
+            assertEquals("foo", promise.get());
+            assertNotEquals(before, started.get());
+        }
+    }
+
+    @Singleton
+    @Requires(property = "spec.name", value = "TaskQueueInterceptorTest")
+    static final class MyTaskQueueInterceptor implements TaskQueueInterceptor {
+        final AtomicInteger started = new AtomicInteger(0);
+
+        @Override
+        public Queue<Runnable> wrapTaskQueue(String groupName, Queue<Runnable> original) {
+            return new AbstractQueue<>() {
+                @Override
+                public Iterator<Runnable> iterator() {
+                    return original.iterator();
+                }
+
+                @Override
+                public int size() {
+                    return original.size();
+                }
+
+                @Override
+                public boolean offer(Runnable runnable) {
+                    return original.offer(() -> {
+                        started.incrementAndGet();
+                        runnable.run();
+                    });
+                }
+
+                @Override
+                public Runnable poll() {
+                    return original.poll();
+                }
+
+                @Override
+                public Runnable peek() {
+                    return original.peek();
+                }
+            };
+        }
+    }
+}

